Suggest method names without checking for capitalization

lizmat opened the pull request:
Currently, an expression like `say 42.which` will fail with:

  No such method 'which' for invocant of type 'Int'

*without* suggesting .WHICH.  This patch changes that to:

  No such method 'which' for invocant of type 'Int'.  Did you mean 'WHICH'?

lizmat committed:
Suggest method names that differ only in capitalization

Currently, an expression like `say 42.which` will fail with:

  No such method 'which' for invocant of type 'Int'

*without* suggesting .WHICH.  This patch changes that to:

  No such method 'which' for invocant of type 'Int'.  Did you mean 'WHICH'?

vrurg commented on src/core.c/Exception.pm6:
```diff
@@ -203,15 +203,26 @@ my class X::Method::NotFound is Exception {
+                }
+                else {
+                    my $dist =
+                      StrDistance.new(:before($.method), :after($method_name));
```
Cancelled an accidental review and lost my comment. So, just repeating myself again: what if we just use `.fc` on both variants here?

```
StrDistance.new(:before($.method.fc), :after($method_name.fc));
```

It would then cover erratas like `.whicj` too.

lizmat replied:
Good point.

lizmat replied:
Added with 30c4b6e599

lizmat committed:
Perform Levenshtein on .fc of strings

To more easily cover capitalization errors

lizmat committed:
Make test less specific

It was breaking because the number of suggestions changed.

lizmat committed:
Add checks for HOW WHAT WHO manually

As suggested as a solution for https://github.com/rakudo/rakudo/pull/3854

codesections commented on t/05-messages/01-errors.t:
```diff
@@ -222,7 +222,7 @@ throws-like { Blob.splice }, X::Multi::NoMatch,
     throws-like q| class RT123078_3 { method !bar { }; method baz { } }; RT123078_3.new.bar |,
         X::Method::NotFound,
-        message => all(/<<"No such method 'bar'" \W/, /<<'RT123078_3'>>/, /\s+ Did \s+ you \s+ mean \s+ "'baz'"/),
+        message => all(/<<"No such method 'bar'" \W/, /<<'RT123078_3'>>/, /\s+ Did \s+ you \s+ mean/),
```
This test was previously testing that "a private method of the same name as the public missing method is not suggested for out-of-class call" – that is, that `!bar` was _not_ suggested.  This revised test no longer ensures that `!bar` is not suggested, which could allow a regression in future code changes.

Here is the alternate test I proposed in my accidental duplicate PR, which I believe addressed that issue. 

```raku
       message => all(/<<"No such method 'bar'" \W/, /<<'RT123078_3'>>/, /<<'baz'>>/,  none(/<<'!bar'>>/)),
```

codesections reviewed (commented):
I noticed one issue where the test could allow a regression; I've described the issue and proposed a solution inline.

lizmat committed:
Externalize suggestions and tips

The X::Method::NotFound error is hard to test for.  This commit
adds the "suggestions" and "tips" method, so that one can test for
(not) expected suggestions, and whether any tips were generated.

Also only use numeric weights in suggestions: there was a mix of
strings and integers, that can't be right  :-)

lizmat committed:
Add test for specific suggestions

vrurg reviewed (approved):
Hopefully, I miss nothing.

vrurg commented:
@lizmat If there is no other reason to keep this opened, I'd say it's ready to merge.
